tests/unit/test_parameterisations.py

```diff
@@ -71,75 +71,108 @@ def spm_cost(self, parameters, model, x0):
 
         # Define the cost to optimise
-        signal = "Terminal voltage [V]"
+        signal = ["Terminal voltage [V]"]
         problem = pybop.FittingProblem(
             model, parameters, dataset, signal=signal, init_soc=init_soc
         )
         return pybop.SumSquaredError(problem)
 
-    @pytest.mark.unit
-    def test_spm_optimisers(self, spm_cost, x0):
-        # Select optimisers
-        optimisers = [
+    @pytest.mark.parametrize(
+        "optimiser",
+        [
             pybop.NLoptOptimize,
             pybop.SciPyMinimize,
             pybop.SciPyDifferentialEvolution,
-            pybop.CMAES,
             pybop.Adam,
+            pybop.CMAES,
             pybop.GradientDescent,
+            pybop.IRPropMin,
             pybop.PSO,
-            pybop.XNES,
             pybop.SNES,
-            pybop.IRPropMin,
-        ]
+            pybop.XNES,
+        ],
+    )
+    @pytest.mark.unit
+    def test_spm_optimisers(self, optimiser, spm_cost, x0):
+        # Test each optimiser
+        parameterisation = pybop.Optimisation(cost=spm_cost, optimiser=optimiser)
+        parameterisation.set_max_unchanged_iterations(iterations=15, threshold=5e-4)
+
+        if optimiser in [pybop.CMAES]:
+            parameterisation.set_f_guessed_tracking(True)
+            assert parameterisation._use_f_guessed is True
+            parameterisation.set_max_iterations(1)
+            x, final_cost = parameterisation.run()
+
+            parameterisation.set_f_guessed_tracking(False)
+            parameterisation.set_max_iterations(100)
+
+            x, final_cost = parameterisation.run()
+            assert parameterisation._max_iterations == 100
+
+        elif optimiser in [pybop.GradientDescent]:
+            parameterisation.optimiser.set_learning_rate(0.025)
+            parameterisation.set_max_iterations(100)
+            x, final_cost = parameterisation.run()
+
+        else:
+            parameterisation.set_max_iterations(100)
+            x, final_cost = parameterisation.run()
+
+        # Assertions
+        np.testing.assert_allclose(final_cost, 0, atol=1e-2)
+        np.testing.assert_allclose(x, x0, atol=1e-1)
+
+    @pytest.fixture
+    def spm_cost_2(self, parameters, model, x0):
+        # Form dataset
+        init_soc = 0.5
+        solution = self.getdata(model, x0, init_soc)
+        dataset = pybop.Dataset(
+            {
+                "Time [s]": solution["Time [s]"].data,
+                "Current function [A]": solution["Current [A]"].data,
+                "Terminal voltage [V]": solution["Terminal voltage [V]"].data,
+            }
+        )
 
+        # Define the cost to optimise
+        signal = ["Terminal voltage [V]", "Time [s]"]
+        problem = pybop.FittingProblem(
+            model, parameters, dataset, signal=signal, init_soc=init_soc
+        )
+        return pybop.SumSquaredError(problem)
+
+    @pytest.mark.parametrize(
+        "optimiser",
+        [
+            pybop.NLoptOptimize,
```

Comment:
Out of curiosity, what was the reason for these four optimisers?

Reply:
For speed of testing, I chose a subset covering the pybop options and the pints optimiser with and without sensitivities.
But I was wondering- which approach do you think we should take:
1. Test all optimisers using `spm_cost` and then test a subset again using `spm_two_signal_cost`, or
2. For speed, test half the optimisers with `spm_cost` and half with `spm_two_signal_cost`?

Reply:
Good shout, that makes sense to me too.

Hmm, I would consider test only a single optimiser for `spm_two_signal_cost`, or maybe just a single of each (pybop/pints). As an aside, what does the optimiser do when we have one of the signals as ["Time [s]"]? It's essentially a non-changing output, so maybe it doesn't change anything..

Reply:
Yeah, it shouldn't change the results, but including a second arbitrary signal here will flag if there is any mismatch in the array shapes. I think the set of 4 optimisers is required to properly check the array shapes are understood by all optimisers.